examples/cvae/baseline.py

```diff
@@ -0,0 +1,117 @@
+        y = torch.sigmoid(self.fc3(hidden))
+        return y
+
+
+class MaskedBCELoss(nn.Module):
+    def __init__(self, reduction='sum', masked_with=-1):
+        super().__init__()
+        self.reduction = reduction
+        self.masked_with = masked_with
+
+    def forward(self, input, target):
+        target = target.view(input.shape)
+        loss = F.binary_cross_entropy(input, target, reduction='none')
+        loss[target == self.masked_with] = 0
+        if self.reduction == 'sum':
+            return loss.sum()
+        elif self.reduction == 'mean':
+            return loss.sum() / (target != -1).sum()
+        else:
+            raise ValueError("{} is not a valid value".format(self.reduction))
+
+
+def train(device, dataloaders, dataset_sizes, learning_rate, num_epochs,
+          early_stop_patience, model_path):
+
+    # Train baseline
+    baseline_net = BaselineNet(500, 500)
+    baseline_net.to(device)
+    optimizer = torch.optim.Adam(baseline_net.parameters(), lr=learning_rate)
+    criterion = MaskedBCELoss(reduction='sum')
+    best_loss = np.inf
+    early_stop_count = 0
+
+    for epoch in range(num_epochs):
+        # Each epoch has a training and validation phase
+        for phase in ['train', 'val']:
+            if phase == 'train':
+                baseline_net.train()  # Set model to training mode
+            else:
+                baseline_net.eval()   # Set model to evaluate mode
+
+            running_loss = 0.0
+            num_preds = 0
+
+            # Iterate over data.
+            bar = tqdm(dataloaders[phase],
+                       desc='NN Epoch {} {}'.format(epoch, phase).ljust(20))
+            for i, batch in enumerate(bar):
+                inputs = batch['input'].to(device)
+                outputs = batch['output'].to(device)
+
+                # zero the parameter gradients
+                optimizer.zero_grad()
+
+                # forward
+                # track history if only in train
+                with torch.set_grad_enabled(phase == 'train'):
+                    preds = baseline_net(inputs)
+                    loss = criterion(preds, outputs) / inputs.size(0)
+
+                    # backward + optimize only if in training phase
+                    if phase == 'train':
+                        loss.backward()
+                        optimizer.step()
+
+                # statistics
+                running_loss += loss.item()
+                num_preds += 1
+                if i % 10 == 0:
+                    bar.set_postfix(loss='{:.2f}'.format(running_loss / num_preds),
+                                    early_stop_count=early_stop_count)
+
+            epoch_loss = running_loss / dataset_sizes[phase]
+            # deep copy the model
+            if phase == 'val':
+                if epoch_loss < best_loss:
+                    best_loss = epoch_loss
+                    best_model_wts = copy.deepcopy(baseline_net.state_dict())
```

Comment:
is this sort-of-complex training mechanism necessary to get decent performance? if not it might be preferable to simplify?

Reply:
This is actually required for the early stopping, which is how the authors trained the models. These `if`'s are, imho, the simplest way to implement early stopping mechanism in PyTorch. They came from this PyTorch tutorial (https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html). 
Personally, I find a bit ugly as well, but I don't know how else to implement early stopping in PyTorch in as few lines as possible... pls advise!

Reply:
i guess i'm asking if you still got ok performance without early stopping, e.g. if you just chose a reasonable number of epochs to train

Reply:
Yes, if I set the number of epochs to a reasonably high value, and different values for the NN and CVAE, yes I can get ok performance without early stopping. The NN usually stops improving around 20 epochs, while the CVAE trains for (way) longer, and if I increase the early stop patience, I could improve the results even further...

Reply:
ok well as far as i'm concerned it's up to you if you want to keep the early stopping code. for simplicity it might be better to remove it but it's up to you